### PR TITLE
[management] secure_backend to dyn storage

### PR DESCRIPTION
**Pull request by davidiw**
This introduces a means to convert from the new command-line storage
parameter to an actual storage backend.

The format for the command-line is
"k0=v0;k1=v1;..."
where at least one of the k_i must be backend and the rest are options
for that backend.

The test vectors show what should be parseable and what is not. Some
advanced parameters, for example, the server address or token for Vault
will only be tested upon use.

### CODE REVIEW

File: config/management/src/secure_backend.rs [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+        let backend = match self.backend.as_ref() {
+            "disk" => {
+                let mut config = OnDiskStorageConfig::default();
+                config.set_data_dir(PathBuf::from(""));
+                let path = self.parameters.remove("path");
+                let path = path.expect("Unable to retrieve path from disk backend");
+                config.path = PathBuf::from(path);
+                SafetyRulesBackend::OnDiskStorage(config)
+            }
+            "memory" => SafetyRulesBackend::InMemoryStorage,
```
Comment (JoshLind) [inline config/management/src/secure_backend.rs:52]: (No change required, but I'd probably put memory first as it's simpler.)

File: config/management/src/secure_backend.rs [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+}
+
+impl Into<SafetyRulesBackend> for SecureBackend {
+    fn into(mut self) -> SafetyRulesBackend {
+        let backend = match self.backend.as_ref() {
+            "disk" => {
+                let mut config = OnDiskStorageConfig::default();
+                config.set_data_dir(PathBuf::from(""));
+                let path = self.parameters.remove("path");
+                let path = path.expect("Unable to retrieve path from disk backend");
```
Comment (JoshLind) [inline config/management/src/secure_backend.rs:48]: Nit: can't we just say "let path = self.parameters.remove("path").expect("Unable to....");"?

File: config/management/src/secure_backend.rs [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+            backend,
+            parameters,
+        }
+    }
+}
+
+impl Into<SafetyRulesBackend> for SecureBackend {
+    fn into(mut self) -> SafetyRulesBackend {
+        let backend = match self.backend.as_ref() {
+            "disk" => {
```
Comment (JoshLind) [inline config/management/src/secure_backend.rs:44]: I'd make these constants, so it's super obvious the strings we're matching on.

File: secure/storage/src/lib.rs [hunk 4]
```diff
@@ -29,5 +32,21 @@ pub use crate::{
     vault::VaultStorage,
 };
 
+impl From<&SafetyRulesBackend> for Box<dyn Storage> {
```
Comment (JoshLind) [inline secure/storage/src/lib.rs:35]: Could we not also update the safety_rules_manager.rs file to use this implementation? I think it currently does this by hand.

File: config/management/src/secure_backend.rs [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+                let mut config = OnDiskStorageConfig::default();
+                config.set_data_dir(PathBuf::from(""));
+                let path = self.parameters.remove("path");
+                let path = path.expect("Unable to retrieve path from disk backend");
+                config.path = PathBuf::from(path);
+                SafetyRulesBackend::OnDiskStorage(config)
+            }
+            "memory" => SafetyRulesBackend::InMemoryStorage,
+            "vault" => {
+                let server = self.parameters.remove("server");
```
Comment (JoshLind) [inline config/management/src/secure_backend.rs:54]: I'd also make these special strings constants, so it's super obvious what strings we're matching on.

File: config/management/src/secure_backend.rs [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+    #[test]
+    #[should_panic]
+    fn test_memory_fail() {
+        let memory = "backend=memory;extra=stuff";
+        storage(memory);
+    }
+
+    #[test]
+    fn test_disk() {
+        let disk = "backend=disk;path=some_path";
```
Comment (JoshLind) [inline config/management/src/secure_backend.rs:103]: Nit: does the implementation currently support whitespace? i.e., what if I wanted to do:

"backend=vault ; server=http://127.0.0.1:8080 ; token=123456 ; namespace=test"

Whitespace just seems to make these long strings a lot more readable.

File: config/management/src/secure_backend.rs [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+// SPDX-License-Identifier: Apache-2.0
+
+use libra_config::config::{OnDiskStorageConfig, SafetyRulesBackend, VaultConfig};
+use libra_secure_storage::Storage;
+use std::{collections::HashMap, convert::Into, path::PathBuf};
+
+/// SecureBackend is a parameter that is stored as a colon separated key/value pairs. The only
+/// expected key is backend which defines which of the SecureBackends the parameters refer to.
+/// Some backends require parameters others do not, so that requires a conversion into the
+/// SafetyRulesBackend type to parse.
```
Comment (JoshLind) [inline config/management/src/secure_backend.rs:11]: Could we not explicitly update the comment to outline what parameters are acceptable? I know this is enforced by the tests, but a simple comment is quicker to read and might save someone a lot of time if they want to use this.
For example:
- For in_memory you only need to state: "backend=memory". No other key/value pairs are required.
- For on_disk, you also need to state the path to the on-disk file, e.g."backend=on_disk ; path=path_to_file".
...

File: config/management/src/secure_backend.rs (hunk 1 above)
Comment (davidiw) [reply config/management/src/secure_backend.rs:52]: alphabetical!

File: config/management/src/secure_backend.rs (hunk 2 above)
Comment (davidiw) [reply config/management/src/secure_backend.rs:48]: that will result in 4 lines versus 2

File: config/management/src/secure_backend.rs (hunk 3 above)
Comment (davidiw) [reply config/management/src/secure_backend.rs:44]: that's fair, I was planning a refactor of this code after our two efforts merged

File: secure/storage/src/lib.rs (hunk 4 above)
Comment (davidiw) [reply secure/storage/src/lib.rs:35]: yup! after our efforts merge, I don't want to make the merging more complicated than it might be...

File: config/management/src/secure_backend.rs (hunk 5 above)
Comment (davidiw) [reply config/management/src/secure_backend.rs:54]: agreed, let's save that for a refactor?

File: config/management/src/secure_backend.rs (hunk 6 above)
Comment (davidiw) [reply config/management/src/secure_backend.rs:103]: no, it does not, worth thinking about the tradeoffs... personally not a big fan of whitespace :P

File: config/management/src/secure_backend.rs (hunk 7 above)
Comment (davidiw) [reply config/management/src/secure_backend.rs:11]: Good call! I plan to do that when I integrate it into structopt

**Comment by davidiw** [comment]
This is the first part. I should have called follow ups:
* Integrate with your PR as there are undoubtedly cleanups that can happen once both PRs land, probably better to hack around that so we minimize dependency conflicts for our efforts
* Push the config -> Box<dyn Storage> all over the place -- for example, it may be useful for KeyManager if / when we support other secure backends
* (Most Important) extend this into the Management and add command-line hints on how to use this

File: config/management/src/secure_backend.rs (hunk 2 above)
Comment (JoshLind) [reply config/management/src/secure_backend.rs:48]: But less characters...  :P (No change required!)

File: config/management/src/secure_backend.rs (hunk 3 above)
Comment (JoshLind) [reply config/management/src/secure_backend.rs:44]: Sounds good!

File: secure/storage/src/lib.rs (hunk 4 above)
Comment (JoshLind) [reply secure/storage/src/lib.rs:35]: Sounds good!

File: config/management/src/secure_backend.rs (hunk 5 above)
Comment (JoshLind) [reply config/management/src/secure_backend.rs:54]: Sounds good!

File: config/management/src/secure_backend.rs (hunk 6 above)
Comment (JoshLind) [reply config/management/src/secure_backend.rs:103]: Okay.. personally I find it somewhat hard to parse, but it's only really in the vault case.

File: config/management/src/secure_backend.rs (hunk 7 above)
Comment (JoshLind) [reply config/management/src/secure_backend.rs:11]: Sounds good!

File: config/management/src/lib.rs [hunk 8]
```diff
@@ -339,58 +392,47 @@ pub mod tests {
...
+        let okey = owner_key(namespace).unwrap();
         assert_eq!(key, okey);
     }
 
+
     #[test]
-    #[should_panic]
     #[ignore]
-    fn test_operator_key_invalid() {
+    fn test_operator_key_valid() {
```
Comment (JoshLind) [inline config/management/src/lib.rs]: nit: call this "test_operator_key" seeing as though it tests both the valid and invalid cases?

File: config/management/src/secure_backend.rs [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+
+use crate::error::Error;
+use libra_config::config::{self, OnDiskStorageConfig, VaultConfig};
+use std::{
+    collections::HashMap,
+    convert::{TryFrom, TryInto},
+    path::PathBuf,
+};
+
+/// SecureBackend is a parameter that is stored as a colon separated key/value pairs. The only
```
Comment (JoshLind) [inline config/management/src/secure_backend.rs:18]: nit: "that is stored as colon separated.."?

File: config/management/src/lib.rs [hunk 10]
```diff
@@ -479,40 +505,52 @@ pub mod tests {
         storage
             .create(constants::WAYPOINT, Value::String("".into()), &policy)
             .unwrap();
-
-        storage
     }
 
-    fn operator_key(namespace: &str) -> Result<Ed25519PublicKey, Error> {
```
Comment (JoshLind) [inline config/management/src/lib.rs]: Can we remove the duplication between "operator_key()" and "owner_key()"? It seems the only difference is the key name, right?

File: config/management/src/lib.rs [hunk 11]
```diff
@@ -287,16 +271,61 @@ impl Command {
...
     #[structopt(long)]
-    token: String,
-    #[structopt(long)]
-    namespace: Option<String>,
+    remote: Option<SecureBackend>,
+}
+
+#[derive(Debug, StructOpt)]
+pub struct SingleBackend {
+    /// The secure backend. Secure backends are represented as a semi-colon
```
Comment (JoshLind) [inline config/management/src/lib.rs:320]: You don't think it's maybe clearer to have this comment above the "SecureBackend" struct, rather than having to duplicate it and place it above all users of the struct?

File: config/management/src/lib.rs (hunk 8 above)
Comment (davidiw) [reply config/management/src/lib.rs]: there's probably some rebase hell going on here, I realized in development that my order of operations wasn't that great. So I went back to the root commit and did some foundational work and rebased on top of it...

File: config/management/src/secure_backend.rs (hunk 9 above)
Comment (davidiw) [reply config/management/src/secure_backend.rs:18]: and it is actually semi-colon lol

File: config/management/src/lib.rs (hunk 10 above)
Comment (davidiw) [reply config/management/src/lib.rs]: look at the top level, is it still excessively duplicated?

File: config/management/src/lib.rs (hunk 11 above)
Comment (davidiw) [reply config/management/src/lib.rs:320]: this is the problem of structopt :/

I imagine there might be some way by looking at how clap works (the library structopt is build on top of)
